src/Storages/MergeTree/MergeTreeBackgroundExecutor.h

```diff
@@ -0,0 +1,144 @@
+ *  Which is a self-written coroutine. It suspends, when returns true from execute() method.
+ *
+ *  Executor consists of ThreadPool to execute pieces of a task (basically calls 'execute' on a task)
+ *  and a scheduler thread, which manages the tasks. Due to bad experience of working with high memory under
+ *  high memory pressure scheduler thread mustn't do any allocations,
+ *  because it will be a fatal error if this thread will die from a random exception.
+ *
+ *  There are two queues of a tasks: pending (main queue for all the tasks) and active (currently executing).
+ *  There is an invariant, that task may occur only in one of these queue. It can occur in both queues only in critical sections.
+ *
+ *  Due to all caveats I described above we use boost::circular_buffer as a container for queues.
+ *
+ *  Another nuisance that we faces with is than background operations always interact with an associated Storage.
+ *  So, when a Storage want to shutdown, it must wait until all its background operaions are finished.
+ */
+class MergeTreeBackgroundExecutor : public shared_ptr_helper<MergeTreeBackgroundExecutor>
+{
+public:
+
+    using CountGetter = std::function<size_t()>;
+    using Callback = std::function<void()>;
+
+    enum class Type
+    {
+        MERGE_MUTATE,
+        FETCH,
+        MOVE
+    };
+
+    MergeTreeBackgroundExecutor(
+        Type type_,
+        CountGetter && threads_count_getter_,
+        CountGetter && max_task_count_getter_,
+        CurrentMetrics::Metric metric_)
+        : type(type_)
+        , threads_count_getter(threads_count_getter_)
+        , max_task_count_getter(max_task_count_getter_)
+        , metric(metric_)
+    {
+        name = toString(type);
+
+        updateConfiguration();
+        scheduler = ThreadFromGlobalPool([this]() { schedulerThreadFunction(); });
+    }
+
+    ~MergeTreeBackgroundExecutor()
+    {
+        wait();
+    }
+
+    bool trySchedule(ExecutableTaskPtr task);
+
+    void removeTasksCorrespondingToStorage(StorageID id);
+
+    void wait();
+
+    size_t activeCount()
+    {
+        std::lock_guard lock(mutex);
+        return active.size();
+    }
+
+    size_t pendingCount()
+    {
+        std::lock_guard lock(mutex);
+        return pending.size();
+    }
+
+private:
+
+    void updateConfiguration();
+
+    static String toString(Type type);
+
+    Type type;
+    String name;
+    CountGetter threads_count_getter;
+    CountGetter max_task_count_getter;
+    CurrentMetrics::Metric metric;
+
+    size_t threads_count{0};
+    size_t max_tasks_count{0};
+
+    AtomicStopwatch update_timer;
+
+    struct Item
+    {
+        explicit Item(ExecutableTaskPtr && task_, CurrentMetrics::Metric metric_)
+            : task(std::move(task_))
+            , increment(std::move(metric_))
+        {
+        }
+
+        ExecutableTaskPtr task;
+        CurrentMetrics::Increment increment;
+        Poco::Event is_done;
```

Comment:
Maybe add atomic_bool here - that task is being waited by storage, and need to be finished beforehand.
We can remove `currently_deleting` then

Reply:
But how we will block the thread which wants to shutdown the storage? We can wait on atomic since c++20, but it will be exactly the same code as we have now.

Reply:
Oh, I see.